src/machine/operation.c

```diff
@@ -154,3 +154,19 @@ Operation *operation_free(Operation *o) {
 
         return mfree(o);
 }
+
+_noreturn_ void report_errno_and_exit(int errno_fd, int r) {
+        assert(errno_fd >= 0);
+
+        if (r < 0) {
+                ssize_t n = write(errno_fd, &r, sizeof(r));
+                if (n < 0)
+                        log_debug_errno(errno, "Failed to write operation's errno: %m");
+                if (n != sizeof(r))
+                        log_debug_errno(SYNTHETIC_ERRNO(EIO), "Sent unexpectedly short message");
+
+                _exit(EXIT_FAILURE);
+        }
+
+        _exit(EXIT_SUCCESS);
+}
```

Comment:
ah, i love this

Reply:
we probably can generalize that eventually, i.e. move to io-util.[ch] and use it at various other places as well.

but let#s leave that for another time.

Reply:
I can do a follow up. Btw, a relevant question: why we don't use and you don't demand using things like `write_loop()/read_loop()` when doing IO?

Reply:
fifos/pipes give you the guarantee that small writes always succeed in full or not at all, according to POSIX, hence write() should be fine. it's mentioned somewhere in the man page.

but yeah, if this is now a bit more generic it might make sense to switch to write_loop() indeed, so that this assumption is not implied, and we also work on sockets and whatnot.